File: problem_54/main.py
from typing import List

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        result: List[int] = []
        row_size = len(matrix)
        col_size = len(matrix[0])

        indexes: List[int] = [0, 0] 
        direction: bool = True
        upper: bool = True
        flag: bool = True
        i: int = 1

        for _ in range(row_size*col_size):
            result.append(matrix[indexes[0]][indexes[1]])

            if upper and (indexes[0] + indexes[1] == col_size-1):
                direction = not direction
                flag = not flag

            if not upper and (indexes[0] + indexes[1] == row_size-1):
                direction = not direction
                flag = not flag

            if indexes[0] + indexes[1] == (row_size-i) + (col_size-i):
                direction = not direction
                upper = not upper
            
            if flag and (indexes[0] - indexes[1] == 1):
                upper = True
                direction = not direction
                i += 1

            if upper:
                indexes[direction] += 1
            else:
                indexes[direction] -= 1

        return result

def main():
    # matrix: List[List[int]] = [[1,2,3],[4,5,6],[7,8,9]]
    matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
    new_matrix = Solution().spiralOrder(matrix)
    print(new_matrix)

if __name__ == '__main__':
    main()
        

File: problem_54/main.cpp
#include <vector>
#include <stdio.h>

using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        
        vector<int> result;
        size_t n = matrix.size();
        size_t p = matrix[0].size();
        if (n == 1 || p == 1) {
            for (size_t i=0; i < n; i++) {
                for (size_t j=0; j<p; j++) {
                    result.push_back(matrix[i][j]);
                }
            }
            return result;
        }
        int row_size = n-1;
        int col_size = p-1;
        vector<int> indexes = {0, 0};
        bool direction = true;
        bool upper = true;
        for (size_t i = 0; i < n*p; i++) {
            result.push_back(matrix[indexes[0]][indexes[1]]);

            if (upper && (indexes[0] + indexes[1]) == col_size) { // row+col=n-1
                direction = !direction;
            }

            if (!upper && (indexes[0] + indexes[1]) == row_size) {
                direction = !direction;
            }

            // if (abs(indexes[0] - indexes[1]) == abs(row_size - col_size)) { // row = col
            if (indexes[0] == row_size && indexes[1] == col_size) { // row = col
                direction = !direction; 
                upper = !upper;
            }
            if (!direction && (indexes[0] - indexes[1] == 1)) {
                // start new spiral
                upper = true;
                direction = !direction;
                row_size--;
                // col_size--;
            }
            if (upper) {
                indexes[direction]++;
            }
            else {
                indexes[direction]--;
            }
        }
        return result;
    }
};

int main() {
    // vector<vector<int>> matrix = {{1,2,3},{4,5,6},{7,8,9}};
    // vector<vector<int>> matrix = {{1,2,3,4},{5,6,7,8},{9,10,11,12}};
    // vector<vector<int>> matrix = {{7},{9},{6}};
    vector<vector<int>> matrix = {{1,2,3,4},{5,6,7,8},{9,10,11,12},{13,14,15,16},{17,18,19,20},{21,22,23,24}};
    Solution solution;
    vector<int> result = solution.spiralOrder(matrix);
    for (int val : result) {
        printf("%d ", val);
    }
    printf("\n");
    return 0;
}